wdl-format/src/format.rs

```diff
@@ -0,0 +1,616 @@
+            pm.format(buffer, state)?;
+            buffer.push_str(NEWLINE);
+        }
+
+        for decl in self.members() {
+            Decl::Unbound(decl).format(buffer, state)?;
+        }
+
+        state.decrement_indent();
+
+        let close_brace = self
+            .syntax()
+            .children_with_tokens()
+            .find(|element| element.kind() == SyntaxKind::CloseBrace)
+            .expect("Struct Definition should have a close brace");
+        format_preceding_comments(&close_brace, buffer, state, false)?;
+        state.indent(buffer)?;
+        buffer.push('}');
+        format_inline_comment(&self.syntax_element(), buffer, state, false)?;
+
+        Ok(())
+    }
+
+    fn syntax_element(&self) -> SyntaxElement {
+        SyntaxElement::Node(self.syntax().clone())
+    }
+}
+
+impl Formattable for DocumentItem {
+    fn format(&self, buffer: &mut String, state: &mut FormatState) -> Result<()> {
+        match self {
+            DocumentItem::Import(_) => {
+                unreachable!("Import statements should not be formatted as a DocumentItem")
+            }
+            DocumentItem::Workflow(workflow) => workflow.format(buffer, state),
+            DocumentItem::Task(task) => task.format(buffer, state),
+            DocumentItem::Struct(structure) => structure.format(buffer, state),
+        }
+    }
+
+    fn syntax_element(&self) -> SyntaxElement {
+        match self {
+            DocumentItem::Import(_) => {
+                unreachable!("Import statements should not be formatted as a DocumentItem")
+            }
+            DocumentItem::Workflow(workflow) => workflow.syntax_element(),
+            DocumentItem::Task(task) => task.syntax_element(),
+            DocumentItem::Struct(structure) => structure.syntax_element(),
+        }
+    }
+}
+
+/// Format a WDL document.
+pub fn format_document(code: &str) -> Result<String, Vec<Diagnostic>> {
+    let (document, diagnostics) = Document::parse(code);
+    if !diagnostics.is_empty() {
+        return Err(diagnostics);
+    }
+    let mut validator = Validator::default();
+    match validator.validate(&document) {
+        std::result::Result::Ok(_) => {
+            // The document is valid, so we can format it.
+        }
+        Err(diagnostics) => return Err(diagnostics),
+    }
+
+    let mut result = String::new();
+    let mut state = FormatState::default();
+
+    let version_statement = document
+        .version_statement()
+        .expect("Document should have a version statement");
+    match version_statement.format(&mut result, &mut state) {
+        Ok(_) => {}
+        Err(_) => {
+            return Err(vec![Diagnostic::error(
+                "Failed to format version statement",
+            )]);
+        }
+    }
+
+    let ast = document.ast();
+    let ast = ast.as_v1().expect("Document should be a v1 document");
+    let mut imports = ast.imports().collect::<Vec<_>>();
```

Comment:
Is `ast.imports()` iterable? If so could we skip the `Vec` allocation and just use it directly?

Reply:
The `Vec` allocation is needed (to my knowledge) in order to sort the imports.

Reply:
Ah, right. You're sorting these.